### PR TITLE
Javascript callbacks

### PR DESCRIPTION
**Pull request by arsher**
Prototype JS callback implementation as discussed in [Issue 825](https://github.com/cefsharp/CefSharp/issues/825).

### CODE REVIEW

File: CefSharp/Internals/JavascriptCallbackSurrogate.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
+﻿using System;
+using System.CodeDom;
+using System.Collections.ObjectModel;
+// Copyright © 2010-2014 The CefSharp Project. All rights reserved.
```
Comment (amaitland) [inline CefSharp/Internals/JavascriptCallbackSurrogate.cs]: Copyright is out of whack.

File: CefSharp.Wpf.Example/ViewModels/BrowserTabViewModel.cs [hunk 2]
```diff
@@ -158,11 +169,13 @@ private void OnWebBrowserLoadError(object sender, LoadErrorEventArgs args)
...
             // Don't display an error for downloaded files where the user aborted the download.
             if (args.ErrorCode == CefErrorCode.Aborted)
+            {
                 return;
+            }
 
             var errorMessage = "<html><body><h2>Failed to load URL " + args.FailedUrl +
-                  " with error " + args.ErrorText + " (" + args.ErrorCode +
-                  ").</h2></body></html>";
+                               " with error " + args.ErrorText + " (" + args.ErrorCode +
```
Comment (amaitland) [inline CefSharp.Wpf.Example/ViewModels/BrowserTabViewModel.cs]: Prefer to keep changes that are purely whitespace out of this for now.

File: CefSharp.BrowserSubprocess/CefRenderProcess.cs [hunk 3]
```diff
@@ -8,20 +8,20 @@
 
 namespace CefSharp.BrowserSubprocess
 {
-    [CallbackBehavior(ConcurrencyMode = ConcurrencyMode.Multiple, IncludeExceptionDetailInFaults=true)]
+    [CallbackBehavior(ConcurrencyMode = ConcurrencyMode.Multiple, IncludeExceptionDetailInFaults = true)]
```
Comment (amaitland) [inline CefSharp.BrowserSubprocess/CefRenderProcess.cs]: Unnecessary whitespace changes.

Prefer if they were reverted.

File: CefSharp.BrowserSubprocess.Core/JavascriptCallbackWrapper.h [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+#include<include\cef_v8.h>
+
+using namespace CefSharp::Internals;
+
+namespace CefSharp
+{
+    namespace Internals
+    {
+        ref class JavascriptCallbackWrapper
```
Comment (amaitland) [inline CefSharp.BrowserSubprocess.Core/JavascriptCallbackWrapper.h]: Prefer if we explicitly set the access modifier. (keeps things consistent)

File: CefSharp.BrowserSubprocess.Core/JavascriptCallbackRegistry.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+#pragma once
+
+#include "JavascriptCallbackWrapper.h"
+
+namespace CefSharp
+{
+    namespace Internals
+    {
+        ref class JavascriptCallbackRegistry
```
Comment (amaitland) [inline CefSharp.BrowserSubprocess.Core/JavascriptCallbackRegistry.h]: Prefer if we explicitly set the access modifier. (keeps things consistent)

File: CefSharp/Internals/JavascriptCallback.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        public JavascriptCallback(long id, int browserId, BrowserProcessServiceHost browserProcess)
+        {
+            this.id = id;
+            this.browserId = browserId;
+            _browserProcess = new WeakReference(browserProcess);
+        }
+
+        private BrowserProcessServiceHost BrowserProcessServiceHost
+        {
+            get { return (BrowserProcessServiceHost)_browserProcess.Target; }
```
Comment (amaitland) [inline CefSharp/Internals/JavascriptCallback.cs]: What happens when `Target = null`? Prob better to use `as`. Also no real need for the property, just cast using `as` inline.

Comment (amaitland) [inline CefSharp/Internals/JavascriptCallback.cs]: Basically I'd prefer if it didn't hide the implementation detail, I think it's important to know that there's a weak reference involved.

File: CefSharp/Internals/JavascriptCallback.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+// Use of this source code is governed by a BSD-style license that can be found in the LICENSE file.
+
+using System;
+using System.Runtime.Serialization;
+using System.Threading.Tasks;
+
+namespace CefSharp.Internals
+{
+    [DataContract]
+    internal sealed class JavascriptCallback : IJavascriptCallback
```
Comment (amaitland) [inline CefSharp/Internals/JavascriptCallback.cs]: To make things a little more `DRY`, should probably inherit `DisposableResource`.

File: CefSharp/Internals/JavascriptCallbackEndpointBehavior.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+using System.Collections.Generic;
+using System.Linq;
+using System.ServiceModel.Channels;
+using System.ServiceModel.Description;
+using System.ServiceModel.Dispatcher;
+
+namespace CefSharp.Internals
+{
+    internal sealed class JavascriptCallbackEndpointBehavior : IEndpointBehavior
```
Comment (amaitland) [inline CefSharp/Internals/JavascriptCallbackEndpointBehavior.cs]: With regards to using an `IEndpointBehavior` to convert to a `DTO`, this becomes inconsistent with the rest of the `JSB` implementation.

Have you used this approach in a production system? Is it rock solid? Any potential memory leaks?

In general I think the idea is interesting, and could possibly be used for all the `WCF` proxy classes. Just need to evaluate of the extra code complexity is worth it?

File: CefSharp3.sln.DotSettings [hunk 9]
```diff
@@ -98,6 +98,7 @@
 	<s:Boolean x:Key="/Default/Environment/SearchAndNavigation/GotoSingleUsageImmediately/@EntryValue">False</s:Boolean>
 	<s:Boolean x:Key="/Default/Environment/SearchAndNavigation/MergeOccurences/@EntryValue">True</s:Boolean>
 	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ESettingsUpgrade_002EMigrateBlankLinesAroundFieldToBlankLinesAroundProperty/@EntryIndexedValue">True</s:Boolean>
+	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ESettingsUpgrade_002EMigrateThisQualifierSettings/@EntryIndexedValue">True</s:Boolean>
```
Comment (amaitland) [inline CefSharp3.sln.DotSettings]: Which particular resharper settings is this for?

**Comment by amaitland** [comment]
I've made a few comments inline, also the `CI` build failed, so that will need to be investigated.

Most of my comments are code style/consistency. Overall my target is to have the project as uniform as possible, which can be a little difficult as style can be such a personal thing. Ideally whitespace changes are reverted at this stage, makes for a much easier code review. If your not familiar then checkout the `Files Changed` tab above you'll have a very easy way to review the changes as a whole.

There's a few things I'd like to discuss further, just need access to `VS` to look into them properly before making further comment.

Overall, fantastic effort! :+1: When I get a chance I'll take it for a proper `test drive`.

File: CefSharp.BrowserSubprocess.Core/JavascriptCallbackRegistry.h [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+        {
+        private:
+            int browserId;
+            Int64 lastId;
+            Dictionary<Int64, JavascriptCallbackWrapper^>^ callbacks = gcnew Dictionary<Int64, JavascriptCallbackWrapper^>();
```
Comment (amaitland) [inline CefSharp.BrowserSubprocess.Core/JavascriptCallbackRegistry.h]: It's likely `callbacks` will need to be initialized in the constructor to avoid the current build errors. (Were you developing on `VS2013`? There are a few differences between that and `VS2012` in terms of `VC++`).

File: CefSharp.BrowserSubprocess.Core/JavascriptCallbackRegistry.h [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+#include "JavascriptCallbackWrapper.h"
+
+namespace CefSharp
+{
+    namespace Internals
+    {
+        ref class JavascriptCallbackRegistry
+        {
+        private:
+            int browserId;
```
Comment (amaitland) [inline CefSharp.BrowserSubprocess.Core/JavascriptCallbackRegistry.h]: `C++` variables are prefixed with `_`.

File: CefSharp.BrowserSubprocess.Core/JavascriptMethodHandler.cpp [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+        delete _callbackRegistry;
+    }
+
+    bool JavascriptMethodHandler::Execute(const CefString& name, CefRefPtr<CefV8Value> object, const CefV8ValueList& arguments, CefRefPtr<CefV8Value>& retval, CefString& exception)
+    {
+        auto parameter = gcnew array<Object^>(arguments.size());
+
+        for (std::vector<CefRefPtr<CefV8Value>>::size_type i = 0; i != arguments.size(); i++)
+        {
+            parameter[i] = TypeUtils::ConvertFromCef(arguments[i]);
```
Comment (amaitland) [inline CefSharp.BrowserSubprocess.Core/JavascriptMethodHandler.cpp]: This line should probably be in an `else` block of the `if` statement below.

File: CefSharp.BrowserSubprocess.Core/JavascriptCallbackRegistry.cpp [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+            {
+                auto callback = _callbacks[id];
+                _callbacks->Remove(id);
+                delete callback;
+            }
+        }
+
+        JavascriptCallbackRegistry::~JavascriptCallbackRegistry()
+        {
+            for each (auto callback in _callbacks)
```
Comment (amaitland) [inline CefSharp.BrowserSubprocess.Core/JavascriptCallbackRegistry.cpp]: Still seeing a build error https://ci.appveyor.com/project/cefsharp/cefsharp/build/39.0.0-CI758#L11316

From memory `auto` in a `for each` can cause problems when building with `2012`, if we explicitly type `callback` then hopefully the problem will go away.

File: CefSharp.BrowserSubprocess.Core/JavascriptCallbackRegistry.h [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+    namespace Internals
+    {
+        private ref class JavascriptCallbackRegistry
+        {
+        private:
+            int _browserId;
+            Int64 _lastId;
+            Dictionary<Int64, JavascriptCallbackWrapper^>^ _callbacks;
+        public:
+            JavascriptCallbackRegistry(int browserId) : _browserId(browserId){
```
Comment (amaitland) [inline CefSharp.BrowserSubprocess.Core/JavascriptCallbackRegistry.h]: Brace on new line.

**Comment by amaitland** [comment]
Excellent, we have a working build :+1:

File: CefSharp.BrowserSubprocess.Core/JavascriptCallbackRegistry.h [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+namespace CefSharp
+{
+    namespace Internals
+    {
+        private ref class JavascriptCallbackRegistry
+        {
+        private:
+            int _browserId;
+            Int64 _lastId;
+            Dictionary<Int64, JavascriptCallbackWrapper^>^ _callbacks;
```
Comment (amaitland) [inline CefSharp.BrowserSubprocess.Core/JavascriptCallbackRegistry.h]: Should we switch this to a `ConcurrentDictionary`?

**Comment by arsher** [comment]
I think I went through all of the minor things, thanks for pointing them out. It seems like there were a few changes in VS2013 around C++ I wasn't aware of, unfortunately I don't do too much C++ dev :).

A few other things (I'm not sure if I should react inline, so forgive me if that's the usual way) :

> What happens when Target = null? Prob better to use as. Also no real need for the property, just cast using as inline.

Casting a null value will be still null, but the property is indeed unnecessary so I removed it.

> Should we switch this to a ConcurrentDictionary?

Most definitely, but I think I'm doing that tomorrow.

> With regards to using an IEndpointBehavior to convert to a DTO, this becomes inconsistent with the rest of the JSB implementation.
> 
> Have you used this approach in a production system? Is it rock solid? Any potential memory leaks?
> 
> In general I think the idea is interesting, and could possibly be used for all the WCF proxy classes. Just need to evaluate of the extra code complexity is worth it?

I used `IDataContractSurrogate` before for similar purposes, although in this case it's only one-way. First I had an implementation without the `DTO` but to be honest I didn't like the fact that the `JavascriptCallback` class had multiple responsibilities and also I still needed a way to pass an instance of `BrowserProcessServiceHost` to it. I know it looks totally different from the other `WCF` DataContract solutions in the project but I think they're all for just transporting data, they don't have to serve as a point of interaction. I'm not sure this is the best solution though, so I'm open to any ideas :). I couldn't think of a way which doesn't seem a bit hacky.

**Comment by amaitland** [comment]
> Casting a null value will be still null, but the property is indeed unnecessary so I removed it.

I couldn't remember, I out of habit use `as`.

> I'm not sure this is the best solution though, so I'm open to any ideas

Let's think about it a little more and see what we can come up with. Perhaps @rassilon has some ideas on the subject?

**Comment by rassilon** [comment]
> > I'm not sure this is the best solution though, so I'm open to any ideas
> Let's think about it a little more and see what we can come up with. Perhaps @rassilon has some ideas on the subject?

Ok, so it looks like the idea of the `IEndPointBehavior/IDataContractSurrogate` is to auto-provide the deseralized JavaScriptCallbacks with the BrowserServiceHost instance. Right?

However, for this particular set of changes....

Couldn't you just inherit from JavaScriptCallback to have a JavaScriptCallbackBrowser and mark the service host as non-seralizable? Isn't there one location in the BrowserProcess to populate the service host into JavaScriptCallbackBrowser other than via `IEndPointBehavior`?

i.e. I don't even recall if you can serialize IJavaScriptCallback via WCF for example. If so, can't you bind JavaScriptCallbackBrowser in the browser process, but JavaScritpCallback in the render process?

In general notes follow:
It certainly seems like a lot of extra reflection overhead for so little gain in this particular instance.
I certainly could understand you might want to do something like this in a non-greenfield scenario, However, this is essentially a new feature to CefSharp so I would tend to argue that such usage of `IEndpointBehavior` would start with -25 hypothetical feature points or so and you'd use it if it outweighs 25 fake feature points or so.

Additionally, I'd ideally like to eventually replace the WCF mechanism with CEF's (i.e. Chromium's) IPC mechanism.

Hopefully, this will help us understand this approach a bit better.

Thanks a lot for your great work here, btw! This is definitely not a simple set of changes.

Super generalist :smile: ,
Bill

**Comment by amaitland** [comment]
> Additionally, I'd ideally like to eventually replace the WCF mechanism with CEF's (i.e. Chromium's) IPC mechanism.

I'd be on board with replacing `WCF` once `39` is out the door :+1:

File: CefSharp.Example/Resources/BindingTest.html [hunk 16]
```diff
@@ -3,106 +3,122 @@
     <head>
         <title>Binding Test</title>
     </head>
-    <body>
-        <p>
-        Result of calling bound.repeat("hi ", 5) =
+<body>
```
Comment (amaitland) [inline CefSharp.Example/Resources/BindingTest.html]: Can the formatting be fixed up? It's hard to actually see what's changed.

**Comment by amaitland** [comment]
> It certainly seems like a lot of extra reflection overhead for so little gain in this particular instance.

I hadn't actually considered the reflection overhead. I think overall it would be best to find a simpler solution.

**Comment by amaitland** [comment]
So I've implemented an alternate to `IDataContractSurrogate` which is here https://github.com/amaitland/CefSharp/commit/746e51432acdca5c6adbda7782418dbcce519281

I'm not sure about this approach either, I'm sure there's a bit more that can be cleaned up. Though I'd put it out there for public review first.

Thoughts?

**Comment by arsher** [comment]
Hi,

let me add a few comments for the above.

> Ok, so it looks like the idea of the IEndPointBehavior/IDataContractSurrogate is to auto-provide the deseralized JavaScriptCallbacks with the BrowserServiceHost instance. Right?

That's right. And additionaly the actual call functionality is not used in the render process so it seemed like a mix of responsibilities using the same object on both sides. Although I agree, it's not how the rest of the WCF stuff designed in CefSharp so probably it should be changed.

> Couldn't you just inherit from JavaScriptCallback to have a JavaScriptCallbackBrowser and mark the service host as non-seralizable? Isn't there one location in the BrowserProcess to populate the service host into JavaScriptCallbackBrowser other than via IEndPointBehavior?
> 
> i.e. I don't even recall if you can serialize IJavaScriptCallback via WCF for example. If so, can't you bind JavaScriptCallbackBrowser in the browser process, but JavaScritpCallback in the render process?

You're right, I think inheritance probably the best way to go if we stick to te `IEndpointBehavior` solution. I've chosen that because this way doesn't matter where we pass the JavascriptCallback object the BrowserProcessService gets injected anyway. 
For example I took quick a look at amaitland's alternative which indeed seems to be nicer, although I think (but it's very much possible I missed something) it won't work if a `JavascriptCallback` is returned as the result of a `EvaluateScriptAsync`. I'm going to try to integrate that change and add some additional support to test this.

> If there's a reason frameId isn't stored here, I would think you'd want to comment about it. I know the non-subprocess stuff doesn't currently expose frameIds, but it should in the future since that's how the V8 context is tracked. (i.e. via BrowserId and frameId)

Actually I store the V8 context for each callback so the browserId is only necessary to lookup the `CefBrowserWrapper` instance in `CefRenderProcess`. Which is again not necesserily the best solution but I'll write this up as a comment too.


So i'll take a closer look at 746e51432acdca5c6adbda7782418dbcce519281 and continue the conversation afterwards.

Cheers,
David

**Comment by amaitland** [comment]
> For example I took quick a look at amaitland's alternative which indeed seems to be nicer, although I think (but it's very much possible I missed something) it won't work if a JavascriptCallback is returned as the result of a EvaluateScriptAsync. I'm going to try to integrate that change and add some additional support to test this.

You are correct, I only got the basic example working, my implementation is partially complete. I wasn't completely happy with the solution, so I though it better to open up the discussion before proceeding.

I've had a bit of a rethink overnight, and if we're planning to replace `WCF` in the near future then most of this is moot. So I'm thinking the way forward is.

- Rename `JavascriptCallback` to `JavascriptCallbackProxy`
- Rename `JavascriptCallbackDto` to `JavascriptCallback`
- Add documentation to `JavascriptCallbackProxy` explaining how `WCF` intercepts
- Change `JavascriptCallbackSurrogate` and `JavascriptCallbackEndpointBehavior` to use a `WeakReference` and pass that around instead of the actual reference.
- Expand example to execute callback in a `Task`, demo the `async` benefit.

**Comment by amaitland** [comment]
Changes made in PR to @arsher's branch

https://github.com/arsher/CefSharp/pull/1

Just needs to add some documentation to the `JavascriptCallbackProxy` class explaining how the `Surrogate` works. After that I'm happy to see this merged. (Ideally we should memory profile the changes).

**Comment by amaitland** [comment]
@arsher Any follow comments?

I can scrape the PR against your code and create one directly against `master` if that's easier for you?

**Comment by amaitland** [comment]
I've created #887 which includes all the changes in this `PR` plus my additional ones.

Feel free to comment there. Would be great to get this merged, this week
